File: scripts/service_client.py
#!/usr/bin/env python3
# encoding: utf-8

import sys
from typing import Final

import rospy

# custom service
from catkin_template.srv import Service, ServiceResponse

# constants
NODE_NAME: Final[str] = 'py_service_client'

SERVICE_NAME: Final[str] = 'service'
SERVICE_TIMEOUT_SECS: Final[float] = 5.0


def main() -> None:

  # initialize ROS Node
  # Note: this step can be ommited, but name resolution will not work
  rospy.init_node(NODE_NAME, sys.argv)

  # wait for service to be available
  rospy.wait_for_service(SERVICE_NAME, rospy.Duration(SERVICE_TIMEOUT_SECS))

  # create proxy to call service 
  service = rospy.ServiceProxy(SERVICE_NAME, Service)

  try:
    # execute service call
    resp: ServiceResponse = service(request = 'py::service::request')
    rospy.loginfo(f"Service call to '{rospy.resolve_name(SERVICE_NAME)} returned response = '{resp.response}'")
  
  except rospy.ServiceException:
    rospy.logerr(f"Failed to call service '{rospy.resolve_name(SERVICE_NAME)}'")


if __name__ == '__main__':
  main()

File: src/service_client.cpp
#include <string>

#include <ros/ros.h> // NodeHandle, ServiceProxy

// custom service
#include <catkin_template/Service.h>  // Service, ServiceRequest, ServiceResponse

// constants
static constexpr auto kNodeName = "cpp_service_client";

static constexpr auto kServiceName = "service";

static constexpr auto kTimeoutSecs = double{5.0};


int main(int argc, char **argv) {
  
  // initialize ROS Node
  ros::init(argc, argv, kNodeName);

  auto handle = ros::NodeHandle();

  // wait for service availability
  ros::service::waitForService(kServiceName, ros::Duration(kTimeoutSecs));

  // create proxy to call service
  ros::ServiceClient service = handle.serviceClient<catkin_template::Service>(kServiceName);

  // create service request and response instances
  auto req = catkin_template::ServiceRequest();
  req.request = "cpp::service::request";

  auto resp = catkin_template::ServiceResponse();

  if (!service.call(req, resp)) {
    ROS_ERROR("Failed to call service '%s'", ros::names::resolve(kServiceName).data());
    return 1;
  }

  ROS_INFO("Service call to '%s' returned response = '%s'", ros::names::resolve(kServiceName).data(), resp.response.data());

  return 0;
}
